### PR TITLE
HPCC-20142 Never return NULL property iterators

### PR DESCRIPTION
**Pull request by timothyklemm**
Return iterator for empty property collections instead of NULL.
Remove lazy property initialization to ensure that possibly empty
property collections exist at the time of each request.

Signed-off-by: Tim Klemm <Tim.Klemm@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by timothyklemm** [comment]
@rpastrana and @RussWhitehead Will you review this change?

File: system/security/shared/SecureUser.hpp [hunk 1]
```diff
@@ -50,6 +50,7 @@ class CSecureUser : implements ISecUser, implements ISecCredentials, public CInt
     CSecureUser(const char *name, const char *pw) : 
         m_name(name), m_pw(pw), m_authenticateStatus(AS_UNKNOWN), m_userID(0), m_status(SecUserStatus_Unknown), m_sessionToken(0)
     {
+        m_parameters.setown(createProperties(false));
```
Comment (rpastrana) [inline system/security/shared/SecureUser.hpp]: are we allowed to do this in the init list?

**Review by rpastrana — COMMENTED** [review]
@timothyklemm a couple of comments

File: system/security/LdapSecurity/ldapsecurity.cpp [hunk 2]
```diff
@@ -25,7 +25,10 @@
 /**********************************************************
  *     CLdapSecUser                                       *
  **********************************************************/
-CLdapSecUser::CLdapSecUser(const char *name, const char *pw) : 
+
+Owned<IProperties> CLdapSecUser::sm_emptyParameters(createProperties(false));
```
[the remark this answers is not in the document]
Comment (timothyklemm) [reply system/security/LdapSecurity/ldapsecurity.cpp]: Comment added.
The advantage of owned is consistency. I chose to not create an example that ignores  the reference counted nature of the object which could then be copied by someone and introduce a memory leak.

File: system/security/shared/SecureUser.hpp (hunk 1 above)
Comment (timothyklemm) [reply system/security/shared/SecureUser.hpp]: Yes. Will make the change in all three constructors.

**Comment by timothyklemm** [comment]
@RussWhitehead and @rpastrana I've updated based on Rodrigo's comments.

**Review by rpastrana — APPROVED** [review]
@timothyklemm

**Comment by richardkchapman** [comment]
@ghalliday Can you sanity-check this

**Comment by ghalliday** [comment]
@timothyklemm looks fine.  Please squash the commits.

**Pull request merged**